provider-example/examples/server.rs

```diff
@@ -0,0 +1,107 @@
+use pki_types::CertificateDer;
+use pki_types::{PrivateKeyDer, PrivatePkcs8KeyDer};
+use rustls::server::Acceptor;
+use rustls::ServerConfig;
+use rustls_provider_example::PROVIDER;
+use std::io::Write;
+use std::sync::Arc;
+
+struct TestPki {
+    server_cert_der: CertificateDer<'static>,
+    server_key_der: PrivateKeyDer<'static>,
+}
+
+impl TestPki {
+    fn new() -> Self {
+        let alg = &rcgen::PKCS_ECDSA_P256_SHA256;
+        let mut ca_params = rcgen::CertificateParams::new(Vec::new());
+        ca_params
+            .distinguished_name
+            .push(rcgen::DnType::OrganizationName, "Rustls Server Acceptor");
+        ca_params
+            .distinguished_name
+            .push(rcgen::DnType::CommonName, "Example CA");
+        ca_params.is_ca = rcgen::IsCa::Ca(rcgen::BasicConstraints::Unconstrained);
+        ca_params.key_usages = vec![
+            rcgen::KeyUsagePurpose::KeyCertSign,
+            rcgen::KeyUsagePurpose::DigitalSignature,
+            rcgen::KeyUsagePurpose::CrlSign,
+        ];
+        ca_params.alg = alg;
+        let ca_cert = rcgen::Certificate::from_params(ca_params).unwrap();
+
+        // Create a server end entity cert issued by the CA.
+        let mut server_ee_params = rcgen::CertificateParams::new(vec!["localhost".to_string()]);
+        server_ee_params.is_ca = rcgen::IsCa::NoCa;
+        server_ee_params.extended_key_usages = vec![rcgen::ExtendedKeyUsagePurpose::ServerAuth];
+        server_ee_params.alg = alg;
+        let server_cert = rcgen::Certificate::from_params(server_ee_params).unwrap();
+        let server_cert_der = CertificateDer::from(
+            server_cert
+                .serialize_der_with_signer(&ca_cert)
+                .unwrap(),
+        );
+        let server_key_der =
+            PrivatePkcs8KeyDer::from(server_cert.serialize_private_key_der()).into();
+        Self {
+            server_cert_der,
+            server_key_der,
+        }
+    }
+
+    fn server_config(self) -> Arc<ServerConfig> {
+        let mut server_config = ServerConfig::builder_with_provider(PROVIDER)
+            .with_safe_defaults()
+            .with_no_client_auth()
+            .with_single_cert(vec![self.server_cert_der], self.server_key_der)
+            .unwrap();
+
+        server_config.key_log = Arc::new(rustls::KeyLogFile::new());
+
+        Arc::new(server_config)
+    }
+}
+
+fn main() {
+    env_logger::init();
+
+    let pki = TestPki::new();
+    let server_config = pki.server_config();
+
+    let listener = std::net::TcpListener::bind(format!("0.0.0.0:{}", 4443)).unwrap();
```

Comment:
Might as well make this workable w/ IPv6
```suggestion
    let listener = std::net::TcpListener::bind(format!("[::]:{}", 4443)).unwrap();
```